### PR TITLE
[api] query staking data by eth_call

### LINKED ISSUE #3681: Web3 staking support query staking data
### Is this a hard-fork (and height to be activated)
- [] Yes
- Name of activation height:

### Implementation plan
- [] Define data struct and view method to staking ABI
- [] Refactor `eth_call` to support query

### Affected modules
- [] action
- [] actpool
- [x] api
- [] blockchain
- [] blockindex
- [] blocksync
- [] chainservice
- [] config
- [] consensus
- [] db
- [] dispatcher
- [] ioctl
- [] p2p
- [] server
- [] state
- [] tools

### PR DESCRIPTION
**Pull request by ququzone**
# Description
Query staking data by `eth_call`

Fixes #3681 

## Type of change
Please delete options that are not relevant.
- [x] New feature (non-breaking change which adds functionality)
- [] Code refactor or improvement
- [] Breaking change (fix or feature that would cause a new or changed behavior of existing functionality)
- [] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [x] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by Liuhaai** [comment]
- Files under `./action` could be moved to `./action/protocol/staking`, b/c they aren't txs to be commited in the chain
 - Encoding for input and output could be separated.

File: action/protocol/staking/ethabi/stake_base.go [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+	data, err := outputs.Pack(args)
+	if err != nil {
+		return "", nil
+	}
+	return hex.EncodeToString(data), nil
+}
+
+func encodeCandidateToEth(candidate *iotextypes.CandidateV2) (*CandidateEth, error) {
+	result := &CandidateEth{}
+	if addr, err := addrutil.IoAddrToEvmAddr(candidate.OwnerAddress); err == nil {
```
Comment (Liuhaai) [inline action/protocol/staking/ethabi/stake_base.go:104]: result.OwnerAddress, err = addrutil.IoAddrToEvmAddr(candidate.OwnerAddress)
if err != nil {
return nil, err
}

File: api/web3server.go [hunk 2]
```diff
@@ -329,6 +330,21 @@ func (svr *web3Handler) call(in *gjson.Result) (interface{}, error) {
 	if to == _metamaskBalanceContractAddr {
 		return nil, nil
 	}
+	if to == address.StakingProtocolAddr {
+		sctx, err := ethabi.CallDataToStakeStateContext(data)
```
Comment (Liuhaai) [inline api/web3server.go:334]: `ethabi.CallDataToStakeStateContext` -> stakingabi.BuildReadStateRequest(...)

**Review by Liuhaai — APPROVED** [review]

File: action/protocol/staking/ethabi/stake_base.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
+package ethabi
+
+import (
+	"encoding/hex"
+	"errors"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/iotexproject/iotex-core/pkg/util/addrutil"
```
Comment (CoderZhi) [inline action/protocol/staking/ethabi/stake_base.go:10]: wrong group

File: action/protocol/staking/ethabi/stake_base.go [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+	Parameters struct {
+		MethodName []byte
+		Arguments  [][]byte
+	}
+
+	// StateContext context for ReadState
+	StateContext interface {
+		Parameters() *Parameters
+		EncodeToEth(*iotexapi.ReadStateResponse) (string, error)
+	}
```
Comment (CoderZhi) [inline action/protocol/staking/ethabi/stake_base.go:33]: it seems that these structs/interfaces are used in multiple protocols. Thus, we may unify them into, say, action/protocol/

**Review by CoderZhi — APPROVED** [review]
some failure cases are not covered in unit test

**Pull request merged**